### DISCUSSION TITLE
How to switch to dev branch?

### DISCUSSION THREAD

**Discussion by eigenphase** [Koenkk/zigbee2mqtt#17938] [General]
Is the info in https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html still current ?

It doesn’t seem so. There is just no dev branch.

`git branch -v -a` just returns master, remotes/origin/HEAD and remotes/origin/master, no matter how often I do `git fetch`, `git fetch origin dev` etc

as a result , `git checkout dev` returns “error: pathspec 'dev' did not match any file(s) known to git”

How can I check out the dev branch (and then switch back to “stable” branch?)

**Comment by henkiejan1** [comment]
It's for me just a mystery how to switch to the dev branch. I got that with git checkout that eveything is up to date but it's just on the normal version...

**Reply by eigenphase** [reply]
Yes this is also my latest status. 

It really seems the official docu is broken…

**Reply by Koenkk** [reply]
can you provide the output of 

'''
git fetch origin
git branch
'''

**Reply by eigenphase** [reply]
```
$ sudo -u zigbee2mqtt git fetch origin
$ sudo -u zigbee2mqtt git branch
  a
* master
$
```

branch “a” might have been a test when I tried to debug this… don’t remember any more. But dev branch doesn’t show up.

**Reply by Koenkk** [reply]
What is the output of `cat .git/config`?

**Reply by eigenphase** [reply]
```
$ cat .git/config
[core]
  repositoryformatversion = 0
  filemode = true
  bare = false
  logallrefupdates = true
[remote "origin"]
  url = https://github.com/Koenkk/zigbee2mqtt.git
  fetch = +refs/heads/master:refs/remotes/origin/master
[branch "master"]
  remote = origin
  merge = refs/heads/master

```

**Reply by Koenkk** [reply]
What if you do `git checkout origin/dev`

**Comment by Nardol** [comment]
Does it work if you type `git fetch origin` so without any branch name?

**Comment by francisp2** [comment]
it is there: 
![image](https://github.com/Koenkk/zigbee2mqtt/assets/52328611/45bed155-1d15-4211-8c6a-1763575e997c)

**Comment by harolds** [comment]
I have the exact same problem. There seems to be no dev branch:

pi@domoticz:/opt/zigbee2mqtt $ git branch -r
  origin/HEAD -> origin/master
  origin/master

And

pi@domoticz:/opt/zigbee2mqtt $ git checkout origin/dev
error: pathspec 'origin/dev' did not match any file(s) known to git

**Reply by SteveBrackenbury** [reply]
Run the following command to view your git fetch config:

`git config --get remote.origin.fetch`

The ouput of the command when run on my workstation revealed that fetch is configured to a specific remote origin branch: master.  The effect being that all other remote branches are effectively "hidden":

`+refs/heads/master:refs/remotes/origin/master`

To make other remote branches visible locally, we need to modify the git fetch config by running the following command which substitues "master" with a wildcard "*" so that **ALL** remote branches will be fetched:

`git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"`

After executing the above git command, run the following command to fetch all remote branches:

`git fetch --all`

You will then see the remote dev branch (and other remote branches) successfully fetched.

You should thenbe able to successfully checkout the dev branch:

```
git checkout dev

M       data/configuration.yaml
branch 'dev' set up to track 'origin/dev'.
Switched to a new branch 'dev'
```

_Originally posted by @SteveBrackenbury in https://github.com/Koenkk/zigbee2mqtt/issues/1174#issuecomment-1874686330_

**Reply by eigenphase** [reply]
Amazing! Thank you!!

I am still surprised that this is nowhere written in the docs...
